src/main/java/com/commercetools/sync/shoppinglists/utils/TextLineItemUpdateActionUtils.java

```diff
@@ -0,0 +1,270 @@
+
+            if (!hasNewTextLineItems) {
+                return emptyList();
+            }
+
+            return newShoppingList.getTextLineItems()
+                                  .stream()
+                                  .filter(Objects::nonNull)
+                                  .map(AddTextLineItemWithAddedAt::of)
+                                  .collect(toList());
+        }
+
+        final List<TextLineItem> oldTextLineItems = oldShoppingList.getTextLineItems();
+        final List<TextLineItemDraft> newTextLineItems = newShoppingList.getTextLineItems()
+                                                                        .stream()
+                                                                        .filter(Objects::nonNull)
+                                                                        .collect(toList());
+
+        return buildUpdateActions(oldShoppingList, newShoppingList, oldTextLineItems, newTextLineItems, syncOptions);
+    }
+
+    /**
+     * The decisions in the calculating update actions are documented on the
+     * `docs/adr/0002-shopping-lists-lineitem-and-textlineitem-update-actions.md`
+     */
+    @Nonnull
+    private static List<UpdateAction<ShoppingList>> buildUpdateActions(
+        @Nonnull final ShoppingList oldShoppingList,
+        @Nonnull final ShoppingListDraft newShoppingList,
+        @Nonnull final List<TextLineItem> oldTextLineItems,
+        @Nonnull final List<TextLineItemDraft> newTextLineItems,
+        @Nonnull final ShoppingListSyncOptions syncOptions) {
+
+        final List<UpdateAction<ShoppingList>> updateActions = new ArrayList<>();
+
+        final int minSize = Math.min(oldTextLineItems.size(), newTextLineItems.size());
+        for (int i = 0; i < minSize; i++) {
+
+            final TextLineItem oldTextLineItem = oldTextLineItems.get(i);
+            final TextLineItemDraft newTextLineItem = newTextLineItems.get(i);
+
+            if (newTextLineItem.getName() == null || newTextLineItem.getName().getLocales().isEmpty()) {
+
+                syncOptions.applyErrorCallback(new SyncException(
+                        format("TextLineItemDraft at position '%d' of the ShoppingListDraft with key '%s' has no name "
+                            + "set. Please make sure all text line items have names.", i, newShoppingList.getKey())),
+                    oldShoppingList, newShoppingList, updateActions);
+
+                return emptyList();
+            }
+
+            updateActions.addAll(buildTextLineItemUpdateActions(
+                oldShoppingList, newShoppingList, oldTextLineItem, newTextLineItem, syncOptions));
+        }
+
+        for (int i = minSize; i < oldTextLineItems.size(); i++) {
+            updateActions.add(RemoveTextLineItem.of(oldTextLineItems.get(i).getId()));
+        }
+
+        for (int i = minSize; i < newTextLineItems.size(); i++) {
+            updateActions.add(AddTextLineItemWithAddedAt.of(newTextLineItems.get(i)));
+        }
```

Comment:
Algorithm here is a bit different from LineItem. Can both be aligned?

Reply:
I will update the ADR.

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/614#issuecomment-728946427